File: dotnet/Lexy/Lexy.Compiler/Parser/ParserLogger.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Lexy.Compiler.Language;
using Microsoft.Extensions.Logging;

namespace Lexy.Compiler.Parser
{
    public class ParserLogger : IParserLogger
    {
        private class LogEntry
        {
            public INode Node { get; }
            public bool IsError { get; }
            public string Message { get; }

            public LogEntry(INode node, bool isError, string message)
            {
                Node = node;
                IsError = isError;
                Message = message;
            }

            public override string ToString() => Message;
        }

        private readonly ILogger<ParserLogger> logger;
        private readonly IList<LogEntry> logEntries = new List<LogEntry>();
        private int failedMessages;
        private IRootNode currentNode;

        public bool HasErrors() => failedMessages > 0;
        public bool HasRootErrors() => logEntries.Any(entry => entry.IsError && entry.Node == null);

        public ParserLogger(ILogger<ParserLogger> logger)
        {
            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
        }

        public void LogInfo(string message) => logger.LogInformation(message);

        public void Log(SourceReference reference, string message)
        {
            if (reference == null) throw new ArgumentNullException(nameof(reference));
            if (message == null) throw new ArgumentNullException(nameof(message));

            logger.LogDebug("{Reference}: {Message}", reference, message);
            logEntries.Add(new LogEntry(currentNode, false, $"{reference}: {message}"));
        }

        public void Fail(SourceReference reference, string message)
        {
            if (reference == null) throw new ArgumentNullException(nameof(reference));
            if (message == null) throw new ArgumentNullException(nameof(message));

            failedMessages++;

            logger.LogError("{Reference}: ERROR - {Message}", reference, message);
            logEntries.Add(new LogEntry(currentNode, true, $"{reference}: ERROR - {message}"));
        }

        public void Fail(INode node, SourceReference reference, string message)
        {
            if (reference == null) throw new ArgumentNullException(nameof(reference));
            if (message == null) throw new ArgumentNullException(nameof(message));

            failedMessages++;

            logger.LogError("{Reference}: ERROR - {Message}", reference, message);
            logEntries.Add(new LogEntry(node, true, $"{reference}: ERROR - {message}"));
        }

        public void LogNodes(IEnumerable<INode> nodes)
        {
            if (!logger.IsEnabled(LogLevel.Debug)) return;

            var nodeLogger = new NodesLogger();
            nodeLogger.Log(nodes);

            logger.LogDebug("Parsed nodes: " + nodeLogger.ToString());
        }

        public bool HasErrorMessage(string expectedError)
        {
            return logEntries.Any(message => message.IsError && message.Message.Contains(expectedError));
        }

        public string FormatMessages()
        {
            return
                $"{string.Join(Environment.NewLine, logEntries)}{Environment.NewLine}";
        }

        public void SetCurrentNode(IRootNode node)
        {
            currentNode = node ?? throw new ArgumentNullException(nameof(node));
        }

        public void Reset() => currentNode = null;

        public bool NodeHasErrors(IRootNode node)
        {
            if (node == null) throw new ArgumentNullException(nameof(node));

            return logEntries.Any(message => message.IsError && message.Node == node);
        }

        public string[] ErrorNodeMessages(IRootNode node)
        {
            return logEntries.Where(entry => entry.IsError && entry.Node == node)
                .Select(entry => entry.Message)
                .ToArray();
        }

        public string[] ErrorRootMessages()
        {
            return logEntries.Where(entry => entry.IsError && entry.Node == null)
                .Select(entry => entry.Message)
                .ToArray();
        }

        public string[] ErrorMessages()
        {
            return logEntries.Where(entry => entry.IsError)
                .Select(entry => entry.Message)
                .ToArray();
        }

        public void AssertNoErrors()
        {
            if (HasErrors())
            {
                throw new InvalidOperationException($"Parsing failed: {FormatMessages()}");
            }
        }
    }
}

File: typescipt/src/parser/parserLogger.ts


export class ParserLogger extends IParserLogger {
   private readonly Array<LogEntry> logEntries = list<LogEntry>(): new;

   private readonly ILogger<ParserLogger> logger;
   private IRootNode currentNode;
   private number failedMessages;

   constructor(logger: ILogger<ParserLogger>) {
     this.logger = logger ?? throw new Error(nameof(logger));
   }

   public hasErrors(): boolean {
     return failedMessages > 0;
   }

   public hasRootErrors(): boolean {
     return logEntries.Any(entry => entry.IsError && entry.Node == null);
   }

   public logInfo(message: string): void {
     logger.LogInformation(message);
   }

   public log(reference: SourceReference, message: string): void {
     if (reference == null) throw new Error(nameof(reference));
     if (message == null) throw new Error(nameof(message));

     logger.LogDebug(`{reference}: {Message}`, reference, message);
     logEntries.Add(new LogEntry(currentNode, false, $`{reference}: {message}`));
   }

   public fail(reference: SourceReference, message: string): void {
     if (reference == null) throw new Error(nameof(reference));
     if (message == null) throw new Error(nameof(message));

     failedMessages++;

     logger.LogError(`{reference}: ERROR - {Message}`, reference, message);
     logEntries.Add(new LogEntry(currentNode, true, $`{reference}: ERROR - {message}`));
   }

   public fail(node: INode, reference: SourceReference, message: string): void {
     if (reference == null) throw new Error(nameof(reference));
     if (message == null) throw new Error(nameof(message));

     failedMessages++;

     logger.LogError(`{reference}: ERROR - {Message}`, reference, message);
     logEntries.Add(new LogEntry(node, true, $`{reference}: ERROR - {message}`));
   }

   public logNodes(nodes: Array<INode>): void {
     if (!logger.IsEnabled(LogLevel.Debug)) return;

     let nodeLogger = new NodesLogger();
     nodeLogger.Log(nodes);

     logger.LogDebug(`Parsed nodes: {Nodes}`, nodeLogger.toString());
   }

   public hasErrorMessage(expectedError: string): boolean {
     return logEntries.Any(message => message.IsError && message.Message.contains(expectedError));
   }

   public formatMessages(): string {
     return
       $`{string.Join(Environment.NewLine, logEntries)}{Environment.NewLine}`;
   }

   public setCurrentNode(node: IRootNode): void {
     currentNode = node ?? throw new Error(nameof(node));
   }

   public reset(): void {
     currentNode = null;
   }

   public nodeHasErrors(node: IRootNode): boolean {
     if (node == null) throw new Error(nameof(node));

     return logEntries.Any(message => message.IsError && message.Node == node);
   }

   public string[errorNodeMessages(node: IRootNode): ] {
     return logEntries.Where(entry => entry.IsError && entry.Node == node)
       .Select(entry => entry.Message)
       .ToArray();
   }

   public string[errorRootMessages(): ] {
     return logEntries.Where(entry => entry.IsError && entry.Node == null)
       .Select(entry => entry.Message)
       .ToArray();
   }

   public string[errorMessages(): ] {
     return logEntries.Where(entry => entry.IsError)
       .Select(entry => entry.Message)
       .ToArray();
   }

   public assertNoErrors(): void {
     if (HasErrors()) throw new Error($`Parsing failed: {FormatMessages()}`);
   }

   private class LogEntry {
     public INode Node
     public boolean IsError
     public string Message

     logEntry(node: INode, isError: boolean, message: string): public {
       Node = node;
       IsError = isError;
       Message = message;
     }

     public override toString(): string {
       return Message;
     }
   }
}
